File: vanity_phone_number_c++/lexicon.cc
#include "lexicon.h"

#include <map>

using namespace std;

struct Node {
  bool isWord;  // indicates whether the path <root->node> is a word
  map<char, Node*> suffixes;
  Node() { isWord = false; }
  ~Node() {
    for (auto x : suffixes) {
      delete x.second;
    }
  }
};

Lexicon::Lexicon() : root{nullptr} {}
Lexicon::~Lexicon() { delete root; }

void Lexicon::add(const string& word) { getNode(word)->isWord = true; }

bool Lexicon::containsPrefix(const string& prefix) const {
  return findNode(prefix) != nullptr;
}

bool Lexicon::containsWord(const string& word) const {
  const Node* found = findNode(word);
  return found != nullptr && found->isWord;
}

Node* Lexicon::findNode(const string& str) const {
  Node* curr = root;
  for (size_t pos = 0; pos < str.size() && curr != nullptr; pos++) {
    curr = (curr->suffixes.find(str[pos]) != curr->suffixes.end())
               ? curr->suffixes[str[pos]]
               : nullptr;
  }

  return curr;
}

Node* Lexicon::getNode(const string& str) {
  Node** currp = &root;

  size_t pos = 0;
  while (true) {
    if (*currp == nullptr) *currp = new Node;
    if (pos == str.size()) break;
    currp = &((*currp)->suffixes[str[pos]]);
    pos++;
  }
  return *currp;
}

File: vanity_phone_number_python/lexicon.py
from typing import Dict


class Node:
    def __init__(self):
        self.isWord = False
        self.suffixes: Dict[str, Node] = {}


class Lexicon:
    def __init__(self):
        self.root: Node = Node()

    def add(self, string: str):
        self.getNode(string).isWord = True

    def containsWord(self, string: str) -> bool:
        found = self.findNode(string)
        return found is not None and found.isWord

    def containsPrefix(self, string: str) -> bool:
        return self.findNode(string) is not None

    def getNode(self, string: str) -> Node:
        curr = self.root
        for pos in range(len(string)):
            try:
                curr = curr.suffixes[string[pos]]
            except KeyError:
                curr.suffixes[string[pos]] = Node()
                curr = curr.suffixes[string[pos]]
        return curr

    def findNode(self, string: str) -> Node:
        curr = self.root
        for pos in range(len(string)):
            curr = curr.suffixes[string[pos]] if string[pos] in curr.suffixes else None
            if curr is None:
                break

        return curr
